### PR TITLE
feat: incremental release of whatsapp, basic apis

### PR DESCRIPTION
**Pull request by richard-yak**
Included in this release

- Adding of basic APIs required by a new channel, such as credentials, get campaigns, and their relevant validators
- Adding models files that were in the previous PR's schema changes.
- Adding DTO mappers for whatsapp related responses
- Adding migration files for supporting models, independent to the whatsapp feature

Upcoming releases
- Basic UI updates to support a new channel
- Campaign updates to support a new channel
- Step by step release for whatsapp, and their relevant backend supporting APIs

### CODE REVIEW

File: backend/src/core/services/domain.service.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+// const logger = loggerWithLabel(module)
+
+const getUserDomain = async (userId: number): Promise<string> => {
+  const user = await User.findOne({
+    where: {
+      id: userId,
+    },
+    attributes: ['emailDomain'],
+  })
+  if (user == null) {
```
Comment (stanleynguyen) [inline backend/src/core/services/domain.service.ts:12]: *`===`

File: backend/src/whatsapp/middlewares/whatsapp.middleware.ts [hunk 2]
```diff
@@ -1,24 +1,37 @@
...
-    logger.info(req, res, credentialService)
+    const userId = req.session?.user?.id
+    const userDomain = await DomainService.getUserDomain(userId)
+    // if empty or user not found
+    if (!userDomain) {
+      logger.error({
+        message: 'User does not belong to a domain',
+        action: 'getTemplates',
+      })
+      res.sendStatus(401)
```
Comment (stanleynguyen) [inline backend/src/whatsapp/middlewares/whatsapp.middleware.ts:30]: Let's have a response body that follows our standard fields of `code` and `message` fields. Come back to this after the refactor is merged in for error responses

File: backend/src/whatsapp/middlewares/whatsapp.middleware.ts [hunk 3]
```diff
@@ -29,6 +42,106 @@ export const InitWhatsappMiddleware = (
...
+      return
+    }
+    if (!credLabel) {
+      logger.error({
+        message: 'Empty credentials label in body',
+        action: 'getTemplates',
+      })
+      res.sendStatus(400)
```
Comment (stanleynguyen) [inline backend/src/whatsapp/middlewares/whatsapp.middleware.ts:70]: Similarly about error reponse

File: backend/src/whatsapp/middlewares/whatsapp.middleware.ts [hunk 4]
```diff
@@ -29,6 +42,106 @@ export const InitWhatsappMiddleware = (
...
+    const domainCreds = await credentialService.getDomainCredential(
+      userDomain,
+      credLabel
+    )
+    if (!domainCreds) {
+      logger.error({
+        message: 'Credentials do not belong to this user',
+        action: 'getTemplates',
+      })
+      res.sendStatus(401)
```
Comment (stanleynguyen) [inline backend/src/whatsapp/middlewares/whatsapp.middleware.ts:83]: Similarly about error response

File: backend/src/whatsapp/middlewares/whatsapp.middleware.ts [hunk 5]
```diff
@@ -29,6 +42,106 @@ export const InitWhatsappMiddleware = (
...
+      return
+    }
+    if (!credLabel) {
+      logger.error({
+        message: 'Empty credentials label in body',
+        action: 'getTemplates',
+      })
+      res.sendStatus(400)
```
Comment (stanleynguyen) [inline backend/src/whatsapp/middlewares/whatsapp.middleware.ts:115]: Similarly about error responses

File: backend/src/whatsapp/models/campaign-whatsapp-template.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    autoIncrement: true,
+  })
+  id!: number
+
+  @ForeignKey(() => Campaign)
+  @Column(DataType.INTEGER)
+  campaignId!: number
+
+  @ForeignKey(() => WhatsappTemplate)
+  @Column(DataType.INTEGER)
```
Comment (stanleynguyen) [inline backend/src/whatsapp/models/campaign-whatsapp-template.ts:27]: INT or BIGINT?

File: backend/src/whatsapp/routes/whatsapp-settings.routes.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { WhatsappMiddleware } from '@whatsapp/middlewares'
+import { Router } from 'express'
+
+export const InitWhatsappSettingsRoute = (
+  whatsappMiddleware: WhatsappMiddleware
+): Router => {
+  const router = Router()
+  router.get('/credentials', whatsappMiddleware.getCredentials)
+  router.get('/templates', whatsappMiddleware.getTemplates)
+  router.get('/phone-numbers', whatsappMiddleware.getPhoneNumbers)
```
Comment (stanleynguyen) [inline backend/src/whatsapp/routes/whatsapp-settings.routes.ts:10]: Where are the request validators?

File: backend/src/whatsapp/services/whatsapp-callback.service.ts [hunk 8]
```diff
@@ -1,10 +1,27 @@
...
+  WhatsappCallbackPayload,
+} from '@shared/clients/whatsapp-client.class/interfaces'
 
 const logger = loggerWithLabel(module)
 
-const parseEvent = async (req: Request): Promise<void> => {
-  logger.info('parsing event message', req)
+const parseEvent = async (req: Request): Promise<WhatsappCallbackPayload> => {
+  const callbackPayload: WhatsappCallbackPayload = req.body
+  // although this is a nested loop, in practice it is only 1 entry
```
Comment (stanleynguyen) [inline backend/src/whatsapp/services/whatsapp-callback.service.ts:12]: This comment is a bit ambiguous, could u elaborate?

File: shared/src/clients/whatsapp-client.class/index.ts [hunk 9]
```diff
@@ -59,4 +64,36 @@ export default class WhatsappClient {
...
   }
+
+  public async getTemplates(wabaId: string): Promise<WhatsappTemplate[]> {
+    const res = await this.request({
+      method: 'get',
+      url: `${wabaId}/message_templates`,
+    }).catch((e) => {
+      // this would mean whatsapp server is down
+      throw new Error(e)
+    })
```
Comment (stanleynguyen) [inline shared/src/clients/whatsapp-client.class/index.ts:75]: Please dont mix `async/await` with `then/catch`

File: shared/src/clients/whatsapp-client.class/index.ts [hunk 10]
```diff
@@ -59,4 +64,36 @@ export default class WhatsappClient {
...
+    // even if whatsapp returned an error, it would still have status code: 200
+    // we must check for the error object instead
+    if (res.error) {
+      throw new Error(res)
```
Comment (stanleynguyen) [inline shared/src/clients/whatsapp-client.class/index.ts:79]: Can the message be more specific than the stringified `res`? Like `Error(res.message)` or something (this is just a shot in the dark as I'm not aware of the structure of this obj

File: shared/src/clients/whatsapp-client.class/index.ts [hunk 11]
```diff
@@ -59,4 +64,36 @@ export default class WhatsappClient {
...
+  public async getPhoneNumbers(wabaId: string): Promise<WhatsappPhoneNumber[]> {
+    const res = await this.request({
+      method: 'get',
+      url: `${wabaId}/phone_numbers`,
+    }).catch((e) => {
+      throw new Error(e)
+    })
+    if (res.error) {
+      throw new Error(res)
+    }
```
Comment (stanleynguyen) [inline shared/src/clients/whatsapp-client.class/index.ts:96]: Same as the bunch of comments above

**Review by stanleynguyen — CHANGES_REQUESTED** [review]

**Comment by tan-zx** [comment]
removing myself from the review — will relook at this after it's clearer what we need to build for SGC is!

**Pull request closed without merging**